Refactor cities flow

gonzaponte opened the pull request:
This PR introduces a major change in the organization of the IC cities structure. Details [here](https://next.ific.uv.es/DocDB/0014/001424/001/Restructuring_IC.pdf) and [here](https://next.ific.uv.es/DocDB/0015/001503/001/Restructuring%20IC%202.pdf).

Since this is a major change in the repository, we will be moving to version 2.0.0 when this PR is merged.

gonzaponte committed:
Add sipms_as_hits

gonzaponte committed:
Cosmetics

gonzaponte committed:
Redesign esmeralda

gonzaponte committed:
Add hits_merger and hits_thresholder to components

gonzaponte committed:
Handle corrected energy in threshold_hits

gonzaponte committed:
Add sophronia

gonzaponte committed:
Refactor compute_and_write_tracks

gonzaponte committed:
Update esmeralda tests

gonzaponte committed:
Update esmeralda config

gonzaponte committed:
Generalize hits_and_kdst_from_files for different groups and nodes

gonzaponte committed:
Redesign beersheba

gonzaponte committed:
Remove unused imports

gonzaponte committed:
Update Beersheba config

gonzaponte committed:
Remove summary-related functionality from beersheba

gonzaponte committed:
Transform hitc into df

gonzaponte committed:
Handle corrected energy in merge_nn_hits

gonzaponte committed:
Replace cluster.empty with cluster.zero. TODO: user cluster.empty

gonzaponte committed:
Remove cdst_from_files

gonzaponte committed:
Update beersheba fixture

gonzaponte committed:
Remove low_th checks in esmeralda tests

gonzaponte committed:
Check also corrected energy in hits_functions_test

gonzaponte committed:
Rename tests

gonzaponte committed:
Parametrize some tests

gonzaponte committed:
Refine assumption on list_of_hits composite

gonzaponte committed:
Remove tests

- test_threshold_hits_with_penthesilea: does not make sense anymore as
penthesilea has been reconfigured into something else

- test_merge_nn_hits_exact: not worth to test exact output as
producing it is too complicated. The cities that use it have a test
for an exact result that can only be fulfilled if this one provides
the proper output

gonzaponte committed:
Add test_merge_nn_hit_does_not_leave_nn_hits

gonzaponte committed:
Add files for tests

gonzaponte committed:
Add fixtures for sophronia

gonzaponte committed:
Add tests for sophronia

gonzaponte committed:
Add fixtures for esmeralda

gonzaponte committed:
Cosmetics

gonzaponte committed:
Sort shortest_paths' output

Without this, the order of the output is not fixed and the functions
that use it are not deterministic

gonzaponte committed:
Rebuild esmeralda tests

gonzaponte committed:
Skip test_find_extrema

gonzaponte committed:
Add and update fixtures for beersheba

gonzaponte committed:
Write thresholded hits in beersheba

gonzaponte committed:
Write filter output to file in beersheba

gonzaponte committed:
Rebuild beersheba tests

gonzaponte committed:
Delete skipped tests

martinperezmaneiro committed:
Update test data

gonzaponte committed:
Add tolerances args to assert function in `assert_tables_equality`

gonzaponte committed:
Increase tolerance in `test_beersheba_exact_result`

gonzaponte committed:
Export hits_corrector to components

gonzaponte committed:
Add hit correction to Esmeralda

gonzaponte committed:
Make hit correction optional in esmeralda

gonzaponte committed:
Clean imports in components

gonzaponte committed:
Write paolina hits instead of highTh hits in esmeralda

gonzaponte committed:
Update esmeralda's exact result test file

gonzaponte committed:
Write hits in Isaura

martinperezmaneiro commented on invisible_cities/cities/components.py:
```diff
@@ -904,33 +887,93 @@ def build_hits(pmap, selector_output, event_number, timestamp):
+
+            slice_zs = (peak.times - s1_t) * units.ns * drift_v
+            slice_es = peak.pmts.sum_over_sensors
+            xys      = sipm_xys[peak.sipms.ids]
```
Line 950 repeated from line 938 (I understand peak sipm ids didn't change between these lines)

martinperezmaneiro commented on invisible_cities/cities/sophronia.py:
```diff
@@ -0,0 +1,222 @@
+        for hit in hitc.hits:
+            corr    = get_coef([hit.X], [hit.Y], [hit.Z], hitc.time)[0]
+            hit.Ec  = hit.E * corr
+            hit.xyz = (hit.X, hit.Y, time_to_Z(hit.Z)) # ugly, but temporary
```
Ugly why? What could change here?

gonzaponte replied:
Before the code was creating new hits. In this new version, we update the existing hits. This line is a bit ugly (IMO) but it will look better when we remove the event model.

martinperezmaneiro commented on invisible_cities/cities/sophronia.py:
```diff
@@ -0,0 +1,222 @@
+    - Compute a set of hits for each slice in the rebinned S2 signal.
+    - If there are more than one hit per slice, share the energy
+      according to the charge recorded in the tracking plane.
+"""
```
Should mention the energy correction the city performs now, to distinguish it from the penthesilea description

martinperezmaneiro commented on invisible_cities/cities/sophronia.py:
City structure is clear and easy to follow

martinperezmaneiro commented on invisible_cities/config/sophronia.conf:
```diff
@@ -0,0 +1,62 @@
+# Penthesilea produces a hit collection starting from PMAPS. Since it
```
Minor change to avoid any confusion, like: Sophronia produces a corrected hit collection starting from PMAPs

martinperezmaneiro commented on invisible_cities/reco/hits_functions.py:
Changes in merge and threshold functions are understandable now that they manage both raw and corrected energy

martinperezmaneiro commented on invisible_cities/cities/esmeralda.py:
```diff
@@ -132,32 +81,23 @@ def esmeralda( files_in        : OneOrManyFiles
+    threshold : float
+        minimum pes for a hit (energy)
+    same_peak : bool
+        whether to reassign NN hits' energy only to the hits from the same peak
```
Add fiducial_r parameter explanation

martinperezmaneiro commented on invisible_cities/cities/esmeralda.py:
```diff
@@ -18,18 +18,15 @@
     - DST/Events      - copy of kdst information from penthesilea
```
'kdst information' (just as it is done later for the esmeralda function)

martinperezmaneiro commented on invisible_cities/cities/esmeralda.py:
I'd change in the initial description: 'This city applies a threshold on hits energy and extracts topology information. The input is sophronia containing corrected hits, kdst global information and mc info.'.
Also, delete the lowTh table from the output 

Besides that, all changes in the city are natural for the new structure and easy to follow

martinperezmaneiro commented on invisible_cities/cities/beersheba.py:
Change initial description (esmeralda --> sophronia), and maybe also mention thresholding of hits before the interpolation.

Other than that, expected and natural changes for this city

martinperezmaneiro commented on invisible_cities/cities/beersheba.py:
```diff
@@ -64,6 +63,31 @@
 from .. core                   import system_of_units as units
 
 
+# Temporary. The removal of the event model will fix this.
+from collections import defaultdict
+def hitc_to_df_(hitc):
+    columns = defaultdict(list)
+    for hit in hitc.hits:
+        columns["event"   ].append(hitc.event)
+        columns["time"    ].append(hitc.time)
+        columns["npeak"   ].append(hit .npeak)
+        columns["Xpeak"   ].append(hit .Xpeak)
+        columns["Ypeak"   ].append(hit .Ypeak)
+        columns["nsipm"   ].append(hit .nsipm)
+        columns["X"       ].append(hit .X)
+        columns["Y"       ].append(hit .Y)
+        columns["Xrms"    ].append(hit .Xrms)
+        columns["Yrms"    ].append(hit .Yrms)
+        columns["Z"       ].append(hit .Z)
+        columns["Q"       ].append(hit .Q)
+        columns["E"       ].append(hit .E)
+        columns["Qc"      ].append(hit .Qc)
+        columns["Ec"      ].append(hit .Ec)
+        columns["track_id"].append(hit .track_id)
+        columns["Ep"      ].append(hit .Ep)
+    return pd.DataFrame(columns)
```
One question, why is this transformation done? I see that the change is temporary until the removal of the event model, but I don't see if there is a reason to do this here and now. Was it to test if some things worked or...?

gonzaponte replied:
The previous Beersheba version worked directly on dataframes, not on HitCollections. However, the thresholding step introduced in the new version works with the latter object.
Because the thresholding is also used in Esmeralda (where the use of HitCollection is too extended to be removed) and the plan is to remove the HitCollection object altogether, I decided to do this transformation to be able to use the existing thresholding implementation in Beersheba.

martinperezmaneiro commented on invisible_cities/cities/sophronia_test.py:
```diff
@@ -0,0 +1,89 @@
+    assert cnt.selection.n_failed == 2
+
+
+@mark.skip(reason = "need a new test file for this")
```
If needed, I could try to search for an event that fulfills this and create a file

gonzaponte replied:
Ugh, I had forgotten about this. I don't think we will encounter any MC event meeting this criteria, perhaps it's better to manufacture it specifically for the test.

gonzaponte replied:
I implemented it differently. Applying a high `q_thr` and checking the correct filter. Will push later.

martinperezmaneiro commented on invisible_cities/cities/esmeralda_test.py:
```diff
@@ -3,380 +3,201 @@
+    hits          = dio.load_dst(path_out,   'CHITS' , 'highTh')
+    filter_output = dio.load_dst(path_out, 'Filters', 'topology_select')
+
+    print(summary.evt_nhits)
```
Remove this print?

martinperezmaneiro commented on invisible_cities/reco/paolina_functions.py:
```diff
@@ -136,7 +136,16 @@ def voxels_from_track_graph(track: Graph) -> List[Voxel]:
+    def voxel_pos(x):
+        return x[0].pos.tolist()
+
+    distances = dict(nx.all_pairs_dijkstra_path_length(track_graph, weight='distance'))
+
+    # sort the output so the result is reproducible
+    distances = { v1 : {v2:d for v2, d in sorted(dmap.items(), key=voxel_pos)}
+                  for v1, dmap in sorted(distances.items(), key=voxel_pos)}
```
I remember that using networkx I had a problem because the order of the output was not fixed, and had to do the same, so great point!

martinperezmaneiro commented on invisible_cities/cities/beersheba_test.py:
A first test here is missing (test_beersheba_runs), to agree with the test structure of the other cities, at least the ones reviewed in this PR

martinperezmaneiro reviewed (commented):
In general, I just have some minor comments or questions. The big picture of this PR and the new structure of the flow is clear reading the code

gonzaponte committed:
Rename isaura hits output /CHITS/lowTh to /DECO/Events

gonzaponte committed:
Store paolina hits in isaura

gonzaponte committed:
Add more hit info in beersheba's output

gonzaponte committed:
Fix beersheba test

gonzaponte committed:
Update test files

gonzaponte committed:
Update isaura test

gonzaponte committed:
Fix beersheba test

jwaiton committed:
Add detector_db explicitly to config files

jwaiton committed:
Remove detector_db being defaulted to 'new'

jwaiton committed:
Add detector_db's argument explicitly in test

gonzaponte committed:
Remove duplicate make_hits in Sophronia

gonzaponte committed:
Make sure esmeralda doesn't drop events

jwaiton committed:
Add test to ensure that detector_db is required

gonzaponte committed:
Make `compute_and_write_tracks_info` a sink

gonzaponte committed:
Adapt esmeralda test

gonzaponte committed:
Replace penthesilea with sophronia in all_cities test

gonzaponte committed:
Rename `some_cities` -> `online_cities`

gonzaponte committed:
Skip `test_esmeralda_exact_result` temporarily

gonzaponte committed:
Make hit correction optional in Sophronia

gonzaponte committed:
Add hit corrections to Beersheba

gonzaponte committed:
Make hit corrections in esmeralda non-optional

gonzaponte committed:
Update test files

gonzaponte committed:
Sort hits before writing them in Esmeralda

The networkx library does not determine the order of the output. This
translates into a random order of the hits table, which is irrelevant
for data analysis, but messes up the comparison between tables.

gonzaponte committed:
Recover test_esmeralda_exact_result

gonzaponte committed:
Implement test_sophronia_filters_events_with_only_nn_hits

gonzaponte committed:
Update test file for Isaura

gonzaponte committed:
Improve Sophronia's docstring

gonzaponte committed:
Improve Esmeralda's docstring

gonzaponte committed:
Clean up unused imports

gonzaponte committed:
Improve Isaura's docstring

gonzaponte committed:
Improve Beersheba's docstring

martinperezmaneiro commented on invisible_cities/cities/isaura.py:
```diff
@@ -98,7 +98,7 @@ def isaura( files_in       : OneOrManyFiles
         write_kdst_table = fl.sink( kdst_from_df_writer(h5out), args= "kdst")
 
-        write_hits = fl.sink(hits_writer(h5out, "DECO", "Events"), args="hits")
+        write_hits = fl.sink(hits_writer(h5out, "DECO", "Events"), args="paolina_hits") # from within compute_tracks
```
Not essential, but now Isaura output file is huge, almost as big as the Beersheba output. I understand the utility for certain analyses, but maybe for some productions is not essential and can be avoided. I don't know if this was already discussed, but one option could be making it optional?

gonzaponte replied:
As discussed offline, we will stick to it and consider reducing the output at a later stage.

martinperezmaneiro reviewed (approved):
Latest commits gave the final shape to the refactor of the cities, the code is clear and understandable. Great work, approved!
